research/2d-decoding/rectangle_detector.py

```diff
@@ -32,35 +34,44 @@ def maxs(items, key=lambda a: a):
             maxItems = [items[i]]
     return maxItems
 
+
 def readImage(filename):
     # Clean up the image to get a nice black and white image
     img = cv2.imread(filename)
-    w,h = img.shape[:2]
+    w, h = img.shape[:2]
     # Turn it into a BW image
     img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
     # Blur filter to remove minor noise
-    blur = cv2.GaussianBlur(img,(3,3),0)
+    blur = cv2.GaussianBlur(img, (3, 3), 0)
 
-    block_size = int(min(w,h)/10)
+    block_size = int(min(w, h)/10)
     # make block size odd
-    block_size = block_size + 1 if block_size % 2 == 0 else block_size 
+    block_size = block_size + 1 if block_size % 2 == 0 else block_size
 
     # Threshold the image adaptively: this is convert to binary image in small chunks, choosing intermediate threshold values based on local statistics (avoids very white parts of the immage destroying everything else)
-    thresholdedImage = cv2.adaptiveThreshold(blur,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,block_size,2) # the block size is 10% of the image min axis, 2 is the constant to subtract from the mean, not sure what it changes
+    # the block size is 10% of the image min axis, 2 is the constant to subtract from the mean, not sure what it changes
+    thresholdedImage = cv2.adaptiveThreshold(
+        blur, 255, cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY, block_size, 2)
+    # Justification of the noisy threshold:
+    #  You may see that in very same coloured regions of a picture, the thresholding might cause some binary noise to arrise
+    #  This happens because there is very low contrast in the region
+    #  This dosen't matter for our usecase since arround the edges of the white zones (where there is a hard black/white boundry), there is enough contrast to make the thresholding give us clear binary data, and the adaptive version means its locally clean with higher acuracy
 
     # Invert the thresholded image because black is a 1 and white is a 0
-    cv2.imwrite("thresholded.png", thresholdedImage) # FIXME: rewrite entire script to use cv2 images instead of PIL images to avoid saving
+    # FIXME: rewrite entire script to use cv2 images instead of PIL images to avoid saving
+    cv2.imwrite("thresholded.png", thresholdedImage)
 
     im = Image.open("thresholded.png").convert("L")
-    w,h = im.size
+    w, h = im.size
     pixels = list(im.getdata())
 
     # reduce them all to 0 or 1 and flipped
     pixels = [0 if pixel > 127 else 1 for pixel in pixels]
 
     pixels = [pixels[i:i+w] for i in range(0, len(pixels), w)]
 
-    minRectSurface = (w*h)*0.3/100 # A rectangle must be at least 0.3% of the image size
+    # A rectangle must be at least 0.3% of the image size
```

Comment:
why? What if I have an 8k image with a bunch of codes?

Reply:
If that use case is needed it can be changed I guess, but you can still store like 250 codes in one picture... that should be enough.